control_plane/attachment_service/src/compute_hook.rs

```diff
@@ -244,9 +244,11 @@ impl ComputeHook {
             3,
             10,
             "Send compute notification",
-            backoff::Cancel::new(cancel.clone(), || NotifyError::ShuttingDown),
+            cancel,
         )
         .await
+        .ok_or_else(|| NotifyError::ShuttingDown)
```

Comment:
On a fatal error (e.g. NotifyError::Fatal on line 243), doesn't this change result in us returning ShuttingDown in all cases?

Reply:
`ok_or_else` here runs the closure activates when `backoff::retry` returns `Option::<Result<???, NotifyError>>::None`, so it's basically just "moving" what we had the `backoff::Cancel` for into this main body, removing the need for closure and solving the "we need a centralized `Cancelled` error type" by using `None` as such.

Reply:
Hmm, I'd rather avoid using None as a synonym for cancellation at this interface -- it requires all the callers to remember to do the .ok_or_none on the output

I'm totally on board with the change to use a cancellation token ref instead of clone, but when I look at these alternatives:
- A) `retry(..., &cancel).ok_or(|| Error::Shutdown)` (proposed in this PR)[
- B) `retry(..., Cancel::new(&cancel, || Error::Shutdown))`

I think option B is less error-prone.  I don't love the `Cancel::new` part particularly, maybe the retry function could just take two separate args, a `&CancellationToken`, and the error  to return on cancellation?

Reply:
> it requires all the callers to remember to do the .ok_or_none on the output

No need to remember; the compiler will enforce this. The more problematic IMHO is the then needed `.and_then(|x| x)` which is needed because `Result::flatten` is (still) unstable (probably something to do with the `identity` function).

This PR does all of the changes, so copypasting will work in future as well.

`None` converges towards `proxy` codebase as well where this "common Cancelled type" was already discovered (we later discussed).

Reply:
I feel like I'm missing something.  Let me ask a more specific question: why is it better to return Option<Result<>>  and call ok_or() to flatten it to a Result<>, rather than returning a Result<> (at the cost of specifying the error variant in an arg to retry()).

> None converges towards proxy codebase as well where this "common Cancelled type" was already discovered (we later discussed).

The status quo in pageserver is that functions return a Result where cancellation is one of the error variants: let's stay consistent until/unless we make an overall decision to go another way.

Reply:
We had a call and the disconnect is that I think the `Option<_>` makes sense for `backoff::retry` not necessarily everywhere.

I think the https://github.com/neondatabase/neon/pull/6618/commits/8607b01952b25b2ba1db1dca53ca9354fc596208 can be viewed as a motivating example.